src/promql/src/planner.rs

```diff
@@ -0,0 +1,627 @@
+    TimeIndexNotFoundSnafu, UnknownTableSnafu, UnsupportedExprSnafu,
+};
+use crate::extension_plan::{InstantManipulate, Millisecond, SeriesNormalize};
+
+#[derive(Default, Debug, Clone)]
+struct PromPlannerContext {
+    // query parameters
+    start: Millisecond,
+    end: Millisecond,
+    interval: Millisecond,
+    lookback_delta: Millisecond,
+
+    // planner states
+    time_index_column: Option<String>,
+    value_columns: Vec<String>,
+}
+
+impl PromPlannerContext {
+    fn from_eval_stmt(stmt: &EvalStmt) -> Self {
+        Self {
+            start: stmt.start.duration_since(UNIX_EPOCH).unwrap().as_millis() as _,
+            end: stmt.end.duration_since(UNIX_EPOCH).unwrap().as_millis() as _,
+            interval: stmt.interval.as_millis() as _,
+            lookback_delta: stmt.lookback_delta.as_millis() as _,
+            ..Default::default()
+        }
+    }
+}
+
+pub struct PromPlanner<S: ContextProvider> {
+    schema_provider: S,
+    ctx: PromPlannerContext,
+}
+
+impl<S: ContextProvider> PromPlanner<S> {
+    pub fn stmt_to_plan(stmt: EvalStmt, schema_provider: S) -> Result<LogicalPlan> {
+        let mut planner = Self {
+            schema_provider,
+            ctx: PromPlannerContext::from_eval_stmt(&stmt),
+        };
+        planner.prom_expr_to_plan(stmt.expr)
+    }
+
+    pub fn prom_expr_to_plan(&mut self, prom_expr: PromExpr) -> Result<LogicalPlan> {
+        let res = match &prom_expr {
+            PromExpr::AggregateExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Aggregate",
+            }
+            .fail()?,
+            PromExpr::UnaryExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Unary Expr",
+            }
+            .fail()?,
+            PromExpr::BinaryExpr { lhs, rhs, .. } => {
+                let _left_input = self.prom_expr_to_plan(*lhs.clone())?;
+                let _right_input = self.prom_expr_to_plan(*rhs.clone())?;
+
+                UnsupportedExprSnafu {
+                    name: "Prom Binary Expr",
+                }
+                .fail()?
+            }
+            PromExpr::ParenExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Paren Expr",
+            }
+            .fail()?,
+            PromExpr::SubqueryExpr { .. } => UnsupportedExprSnafu {
+                name: "Prom Subquery",
+            }
+            .fail()?,
+            PromExpr::NumberLiteral { .. } => UnsupportedExprSnafu {
+                name: "Prom Aggregate",
+            }
+            .fail()?,
+            PromExpr::StringLiteral { .. } => UnsupportedExprSnafu {
+                name: "Prom Aggregate",
+            }
+            .fail()?,
+            PromExpr::VectorSelector {
+                name,
+                offset,
+                start_or_end: _,
+                label_matchers,
+            } => {
+                // This `name` should not be optional
+                let name = name.as_ref().unwrap().clone();
```

Comment:
I'm not sure whether this influences the parser, or if an empty metric is allowed in PromQL. Maybe we need further investigation in the follow-up issues. cc @yuanbohan

Reply:
actually:
1. the metric name is optional. like:  `{__name__=~"job:.*"}`
2. the metric name will be in the label matchers, both in Go and Rust Parser
[yacc](https://github.com/prometheus/prometheus/blob/main/promql/parser/generated_parser.y#L507)
[go parser](https://github.com/prometheus/prometheus/blob/main/promql/parser/generated_parser.y#L507)

Reply:
Thanks for this info, I noticed the second point, but I used to think it's a redundant one.

 So the "name" field only exists when the metric is provided in literal formats, like `{http_records}`. Or user can choose not to do this, but rather give a selector only consisting of matches, like `{__name__=~"job:.*", label="a"}`?

Reply:
>  it's a redundant one
Yes, I have updated it.

> So the "name" field only exists when the metric is provided in literal formats, like {http_records}
like `http_records`

you can always trust this:
If metric name exist in user query, it will always exist in label matchers `__name__`

Reply:
>If metric name exist in user query, it will always exist in label matchers __name__

That's good news. I plan to change this to:
- just ignore the name field in selector
- find metric name from matchers
- only support exact match of metric name in this stage.

And we can still keep the parser the same as Go version. In this metric name problem we are expected to be the same as Go version when all four kinds of match rules in `__name__` are supported.